### PR TITLE
Rework the python bindings using conversion traits from arrow-rs 

### PR DESCRIPTION
**Pull request by kszucs**
Alternative to https://github.com/apache/arrow-datafusion/pull/856
Inspired by https://github.com/apache/arrow-datafusion/pull/856#discussion_r686931737
Depends on https://github.com/apache/arrow-rs/pull/691

The CI will probably fail since I hardcoded `arrow-rs` local dependency.

### CODE REVIEW

File: datafusion/src/lib.rs
Comment (cpcloud) [inline datafusion/src/lib.rs:234]: Does the whole module really need to be public?

File: datafusion/Cargo.toml
Comment (cpcloud) [inline datafusion/Cargo.toml:51]: Path-only dependencies are only fine in a private monorepo. Otherwise, you need both a version and a path.

File: python/Cargo.toml
Comment (cpcloud) [inline python/Cargo.toml:35]: same here, this you should avoid a path-only dependency here, but it looks like there might be other considerations I'm unaware of given the use of git.

File: python/Cargo.toml
Comment (cpcloud) [inline python/Cargo.toml:37]: This isn't a public thing that anyone was referring to by name?

File: datafusion/Cargo.toml
Comment (cpcloud) [inline datafusion/Cargo.toml:71]: Are you now using something from `libc`? I didn't catch that in the first pass.

**Comment by cpcloud** [comment]
Any chance of rewording the commits to be something other than single characters?

File: datafusion/src/lib.rs
Comment (kszucs) [reply datafusion/src/lib.rs:234]: I don't think so, will update.

File: datafusion/Cargo.toml
Comment (kszucs) [reply datafusion/Cargo.toml:51]: Only used for local prototyping. It also depends on an arrow-rs PR so need to wait that first.

File: python/Cargo.toml
Comment (kszucs) [reply python/Cargo.toml:35]: I updated both arrow-rs and datafusion which are dependencies of this change, this is a quazy "editable mode".

File: python/Cargo.toml
Comment (kszucs) [reply python/Cargo.toml:37]: Nope, it produces a `datafusion.internals` submodule (which we can also rename using a leading underscore).

File: datafusion/Cargo.toml
Comment (kszucs) [reply datafusion/Cargo.toml:71]: Used for the C data interface between rust and pyarrow.

**Comment by kszucs** [comment]
> Any chance of rewording the commits to be something other than single characters?

Squashed.

**Comment by kszucs** [comment]
@jorgecarleitao @andygrove we should roll up https://github.com/apache/arrow-rs/pull/691 and this PR so we can proceed with the ibis datafusion backend.

File: python/Cargo.toml
Comment (houqp) [reply python/Cargo.toml:34]: Using path here is fine, we do the same thing for ballista as well. Ideally, if the crate needs to be published on crates.io, we will use both `path` and `version` so it will work with `cargo publish`. But in this case, version is not required since we don't have plan to publish the python binding crate to crates.io.

**Comment by houqp** [comment]
The simplification looks great, thanks @kszucs! Looks like there are some minor rust build errors that need to be addressed.

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs
Comment (kszucs) [inline ballista/rust/core/src/serde/logical_plan/to_proto.rs:410]: Just a temporary workaround, but need to add the Map type to the proto definitions.

**Comment by kszucs** [comment]
> The simplification looks great, thanks @kszucs! Looks like there are some minor rust build errors that need to be addressed.

I had to bump the arrow-rs dependency everywhere so we need to address a couple more issues. 

@houqp there is an [async sort order issue](https://github.com/apache/arrow-datafusion/pull/873/checks?check_run_id=3543390034#step:4:1413) which is rather hard to comprehend, any ideas what could have caused this error?

File: python/src/catalog.rs
Comment (jorgecarleitao) [inline python/src/catalog.rs:41]: is this related to `pa.Table`? It is the Python-equivalent of the `TableProvider`, right?

File: python/src/catalog.rs
Comment (jorgecarleitao) [inline python/src/catalog.rs:36]: same here; this is `SchemaProvider` in DataFusion, right?

File: datafusion/src/pyarrow.rs
Comment (jorgecarleitao) [inline datafusion/src/pyarrow.rs:45]: sweet :)

**Review by jorgecarleitao — APPROVED** [review]
Looks great so far! It is a awesome simplification!

I am curious about the naming equivalents but I think we should merge this regardless.

An observation: from here on we are bound to use the same `pyo3` version used by `arrow-rs`.

I suggest an extra reviewer (e.g.  @alamb, @houqp) -- I focused mostly on the Python part.

File: python/src/catalog.rs
Comment (kszucs) [reply python/src/catalog.rs:41]: Nope, it was an attempt to expose the catalog information to ibis, but I may defer the catalog.rs to a follow-up PR instead.

File: python/src/catalog.rs
Comment (kszucs) [reply python/src/catalog.rs:36]: Right. I'm going to remove this module from the PR to avoid confusion.

**Comment by kszucs** [comment]
I need to cover the newly exposed python API with unittests.

**Comment by houqp** [comment]
@kszucs I took a quick look at the `physical_plan::sort_preserving_merge::tests::test_async` error, it is caused by the arrow-rs upgrade. I haven't had the time to root cause it yet. Will probably take a closer look at it today or tomorrow.

**Comment by alamb** [comment]
> @kszucs I took a quick look at the physical_plan::sort_preserving_merge::tests::test_async error, it is caused by the arrow-rs upgrade. I haven't had the time to trace down on the offending arrow-rs commit yet. Will probably take a closer look at it today or tomorrow.

I am pretty sure I saw a failure in the `sort_preserving_merge` tests related to https://github.com/apache/arrow-rs/pull/656

**Comment by kszucs** [comment]
> > @kszucs I took a quick look at the physical_plan::sort_preserving_merge::tests::test_async error, it is caused by the arrow-rs upgrade. I haven't had the time to trace down on the offending arrow-rs commit yet. Will probably take a closer look at it today or tomorrow.
> 
> I am pretty sure I saw a failure in the `sort_preserving_merge` tests related to [apache/arrow-rs#656](https://github.com/apache/arrow-rs/pull/656)

Thanks Andrew for the heads up! I wonder, wouldn't it better if we would assert on record batch equality rather than its string representation?

**Comment by alamb** [comment]
> Thanks Andrew for the heads up! I wonder, wouldn't it better if we would assert on record batch equality rather than its string representation?

@kszucs  -- I agree that doing so would avoid such "small floating point changes require many test changes" type problem.  I think the downside is that (in my opinion) the tests are then harder to read and update. 

Let me prepare a draft PR for updating datafusion to the latest arrow-rs so we can at least decouple the "changes needed for just arrow-rs upgrade" from the other changes in this PR

**Comment by kszucs** [comment]
> @kszucs -- I agree that doing so would avoid such "small floating point changes require many test changes" type problem. I think the downside is that (in my opinion) the tests are then harder to read and update.

I understand, though the test actually tests that the two representations are equal rather than that the results are equal. 
Aren't there other assertion crates providing better checks, perhaps with optional additional context (as a possible future improvement)? 
 
> Let me prepare a draft PR for updating datafusion to the latest arrow-rs so we can at least decouple the "changes needed for just arrow-rs upgrade" from the other changes in this PR

Great, thank You!

**Comment by alamb** [comment]
> I understand, though the test actually tests that the two representations are equal rather than that the results are equal.
Aren't there other assertion crates providing better checks, perhaps with optional additional context (as a possible future improvement)?

🤔  Yes , you are correct. There may be something deeper going on here and perhaps related to the change to use unstable sort in https://github.com/apache/arrow-rs/pull/552. All the more reason to debug it now. I will work on that

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/pull/984 is the PR to update to arrow-rs master. The `sort_preserving_merge` test is indeed failing and I will continue to debug on that PR.

**Comment by houqp** [comment]
@kszucs @jorgecarleitao quick question, do we want to release this change as part of 0.3.0 or delay to 0.4.0?

**Comment by houqp** [comment]
looks like we won't be able to merge this PR any time soon because it depends on arrow-rs 6.x? If that's the case, we would need to schedule it for 0.4.0 release. I wish there is an easy way to do on demand major version release in arrow-rs :)

**Comment by alamb** [comment]
> I wish there is an easy way to do on demand major version release in arrow-rs :)

I think it is only convention preventing releasing Arrow 6 out of lockstep with other languages. If the community wants another major release sooner, the mechanics are straightforward and I am happy to do so.  I don't think I have the time to build consensus though :)

**Comment by houqp** [comment]
I will leave that decision to @kszucs if he wants to include this great improvement in the upcoming 0.3.0 release :) In the long run, I think it's inevitable for us to break the lockstep, i.e. fully decouple the version, with other languages. Because this will certainly happen for the arrow2 crate ;)

**Comment by kszucs** [comment]
It's not urgent to merge this PR, I'll keep it rebased periodically. 

Seems like there is some discussion around independent release cycles on the mailing list, we may bring it up there.

**Comment by houqp** [comment]
@kszucs arrow 6 has now been pulled into datafusion master, you should be unblocked on this now. I can devote some of my time on this to help resolve the conflict too. Let me know if you need any help from me so we can release the new python binding to pypi after this PR is merged :)

File: python/Cargo.toml
Comment (kszucs) [reply python/Cargo.toml]: Renamed it to `_internal`

**Comment by kszucs** [comment]
@houqp this should be ready now. Breaks the previous API at multiple places, but I think that's fine considering the early stage of the datafusion lib. We can also bump the major version.

**Comment by kszucs** [comment]
Cc @cpcloud if you can have a look

**Review by houqp — APPROVED** [review]
LGTM! Thank you @kszucs . I will give @cpcloud @jorgecarleitao @alamb some extra time to review before merging it this weekend or monday.

**Comment by houqp** [comment]
@kszucs do you mind resolving the conflict caused by https://github.com/apache/arrow-datafusion/commit/6b964cbfedc3ee5b08dea8a6c598de7844b95e91? I think  we should merge this PR after that the conflict is resolved and address follow up reviews as separate PRs if there is any.

**Comment by kszucs** [comment]
@houqp rebased

**Review by houqp — APPROVED** [review]

**Comment by houqp** [comment]
Thank you @kszucs ! Great simplification to the python binding :)

**Pull request merged**
